### PR TITLE
8301763: Adding children to wrong index leaves inconsistent state in Parent#childrenSet 

### PR DESCRIPTION
**Pull request by lukostyra**
This issue happened because `childSet` member of Parent was modified during `onProposedChange()` call - that call did not recognize negative indexes as invalid, which caused an exception when actually adding the Node to a List.

This seemed like the simplest solution which doesn't rework a lot of code underneath. Exceptions coming from a backing list/collection technically are handled by `VetoableListDecorator`'s try-catch clauses, however `VetoableListDecorator` does not provide an interface to react when such an exception happens - without it we cannot revert `childSet` back to its original state.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301763](https://bugs.openjdk.org/browse/JDK-8301763): Adding children to wrong index leaves inconsistent state in Parent#childrenSet

### Reviewers
 * [John Hendrikx](https://openjdk.org/census#jhendrikx) (@hjohn - Committer)
 * [Kevin Rushforth](https://openjdk.org/census#kcr) (@kevinrushforth - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1136/head:pull/1136` \
`$ git checkout pull/1136`

Update a local copy of the PR: \
`$ git checkout pull/1136` \
`$ git pull https://git.openjdk.org/jfx.git pull/1136/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1136`

View PR using the GUI difftool: \
`$ git pr show -t 1136`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1136.diff">https://git.openjdk.org/jfx/pull/1136.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1136#issuecomment-1547809277)

### CODE REVIEW

**Comment by hjohn** [comment]
I think it may be better to do this check in the callers of `onProposedChange` (using the assertions in `Objects` to test for valid indices.).  This will also allow to check when the index is too large, which is a similar problem. Note that you can't do the "too large" check in `onProposedChange` as for `remove`, `size()` would be too large, while for `add(int, E)` `size()` would still be valid.

Also, for the cases where `onProposedChange` is called with multiple sets of indices, those don't need a check (they come from calls like `removeAll`, which will all be valid).

Furthermore, it should be an `IndexOutOfBoundsException` as this is specified by the `List` interface.

I noticed there are similar problems in `ObservableListWrapper`.  This method for example:

    @Override
    protected void doAdd(int index, E element) {
        if (elementObserver != null)
            elementObserver.attachListener(element);
        backingList.add(index, element);
    }

This will attach a listener to the element (assuming it is an `Observable`) even if `backingList.add` fails...

**Comment by lukostyra** [comment]
Thanks for your notes, I was wondering if there was a better approach. I'll rework the patch.

**Comment by lukostyra** [comment]
As a small note, the issue is solved with current patches, however I think the best solution would be what I mentioned in the original PR message - an interface in `VetoableListDecorator` which would let us "roll-back" any changes from `onProposedChange` (ex. revert `childSet` to original state) when backing lists throw an exception. This would require a more thorough analysis of what bits use `VetoableListDecorator` and if any roll-backs are necessary there in addition to Parent code.

File: modules/javafx.base/src/main/java/com/sun/javafx/collections/VetoableListDecorator.java [hunk 1]
```diff
@@ -110,6 +112,7 @@ public boolean setAll(E... elements) {
 
     @Override
     public boolean setAll(Collection<? extends E> col) {
+        Objects.requireNonNull(col);
         onProposedChange(Collections.unmodifiableList(new ArrayList<>(col)), 0, size());
```
Comment (hjohn) [inline modules/javafx.base/src/main/java/com/sun/javafx/collections/VetoableListDecorator.java:116]: It's good to make it explicit.  It already threw NPE because of `new ArrayList<>(col)`, so this is not a change.

File: modules/javafx.base/src/main/java/com/sun/javafx/collections/VetoableListDecorator.java [hunk 2]
```diff
@@ -260,6 +267,7 @@ public boolean addAll(int index, Collection<? extends E> c) {
 
     @Override
     public boolean removeAll(Collection<?> c) {
+        Objects.requireNonNull(c);
```
Comment (hjohn) [inline modules/javafx.base/src/main/java/com/sun/javafx/collections/VetoableListDecorator.java:270]: Looks like a good change, if `c` was `null` and the backing list was empty, this would not throw NPE, but now it does.

File: modules/javafx.graphics/src/test/java/test/javafx/scene/ParentTest.java [hunk 3]
```diff
@@ -838,6 +838,60 @@ public void testPickingChildNodeWithDirtyViewOrder() {
...
+        assertThrows(IndexOutOfBoundsException.class, () -> g.getChildren().remove(-1));
+        assertEquals(preTestSize, g.getChildren().size());
+
+        assertThrows(IndexOutOfBoundsException.class, () -> g.getChildren().remove(g.getChildren().size()));
+        assertEquals(preTestSize, g.getChildren().size());
+
+        // below call should throw no exception - if it does, internal state is corrupted
+        g.getChildren().remove(0);
+    }
+
```
Comment (hjohn) [inline modules/javafx.graphics/src/test/java/test/javafx/scene/ParentTest.java:894]: I think you should also add tests for the `null` cases, if they aren't part of this test already.

**Review by hjohn — CHANGES_REQUESTED** [review]
I think `ObservableListWrapper#remove(int, int)` also needs a check index check, or it may end up remove some items and then throw an exception (exceptions thrown from collections should not modify its state).

**Comment by hjohn** [comment]
> As a small note, the issue is solved with current patches, however I think the best solution would be what I mentioned in the original PR message - an interface in `VetoableListDecorator` which would let us "roll-back" any changes from `onProposedChange` (ex. revert `childSet` to original state) when backing lists throw an exception. This would require a more thorough analysis of what bits use `VetoableListDecorator` and if any roll-backs are necessary there in addition to Parent code.

I'm not convinced this is needed or desirable. The `VetoableListDecorator` is currently set up that its owner can already make "preparations" for the change when it knows it won't veto it (there can be only one vetoer, making this reasonable).  Furthermore, it has the wrapped list under its own control, so it knows what to expect (no exceptions can occur (after this fix) if it decides not to veto).

A roll back mechanism would just complicate the code for very little benefit (a correct FX application won't run into these situations, you'll only be likely to encounter them during development). The whole mechanism surrounding the children list and bad additions should be seen as a best effort early warning system (like ConcurrentModificationExeption).

With this fix I don't see any other ways you could get it in a bad state.

**Review by hjohn — APPROVED** [review]

**Review by kevinrushforth — APPROVED** [review]
Looks good.
